src/main/scala/elevate/rise/rules/algorithmic.scala

```diff
@@ -81,30 +120,77 @@ object algorithmic {
       case _              => Failure(createTransposePair)
     }
-    override def toString = "createTransposePair"
+    override def toString: String = "createTransposePair"
   }
 
-  def `_-> T >> T`: Strategy[Rise] = idAfter `;` createTransposePair
-
   def removeTransposePair: Strategy[Rise] = `T >> T -> `
   case object `T >> T -> ` extends Strategy[Rise]  {
     def apply(e: Rise): RewriteResult[Rise] = e match {
       case App(Transpose(), App(Transpose(), x)) => Success(x :: e.t)
       case _                                     => Failure(removeTransposePair)
     }
-    override def toString = "createTransposePair"
+    override def toString: String = "createTransposePair"
   }
 
   // slideSeq fusion
   import rise.OpenCL.primitives._
   import rise.OpenCL.TypedDSL._
 
-  def slideSeqFusion: Strategy[Rise] = `slideSeq(f) >> map(g) -> slideSeq(f >> g)`
-  def `slideSeq(f) >> map(g) -> slideSeq(f >> g)`: Strategy[Rise] = {
-    case expr@App(App(Map(), g), App(App(App(DepApp(DepApp(SlideSeq(rot), sz: Nat), sp: Nat), wr), f), e)) =>
+  def slideSeqFusion: Strategy[Rise] = {
+    case expr@App(App(Map(), g),
+    App(App(App(DepApp(DepApp(SlideSeq(rot), sz: Nat), sp: Nat), wr), f), e)) =>
       Success(slideSeq(rot)(sz)(sp)(wr)(typed(f) >> g)(e) :: expr.t)
-    case expr@App(App(Map(), g), App(App(App(DepApp(DepApp(DepApp(OclSlideSeq(rot), a: AddressSpace), sz: Nat), sp: Nat), wr), f), e)) =>
+    case expr@App(App(Map(), g),
+    App(App(App(DepApp(DepApp(DepApp(OclSlideSeq(rot), a: AddressSpace),
+    sz: Nat), sp: Nat), wr), f), e)) =>
       Success(oclSlideSeq(rot)(a)(sz)(sp)(wr)(typed(f) >> g)(e) :: expr.t)
     case _ => Failure(slideSeqFusion)
   }
+
+  // the inner strategies shouldn't be accessible from the outside
+  // because they might change the semantics of a program
+  case object freshLambdaIdentifier extends Strategy[Rise] {
+    case object freshIdentifier extends Strategy[Rise] {
+      def apply(e: Rise): RewriteResult[Rise] = e match {
+        case Identifier(name) :: t =>
+          Success(Identifier(freshName("fresh_"+ name))(t))
+        case _ => Failure(freshIdentifier)
+      }
+    }
+
+    case class replaceIdentifier(curr: Identifier, newId: Identifier)
+      extends Strategy[Rise] {
+      def apply(e: Rise): RewriteResult[Rise] = e match {
+        case x: Identifier if curr == x => Success(newId)
+        case _ => Failure(replaceIdentifier(curr, newId))
+      }
+    }
+
+    def apply(e: Rise): RewriteResult[Rise] = e match {
+      case Lambda(x,e) :: t if contains[Rise](x).apply(e) => {
+        val newX = freshIdentifier(x).get.asInstanceOf[Identifier]
+        val newE = tryAll(replaceIdentifier(x, newX)).apply(e).get
+        Success(Lambda(newX, newE)(t))
+      }
+      case _ => Failure(freshLambdaIdentifier)
+    }
+    override def toString: String = "freshLambdaIdentifier"
+  }
+
+  case class blockedReduce(n: Nat) extends Strategy[Rise] {
```

Comment:
high-level specification? I think having to read the `apply` code to understand a strategy is not desirable.

Reply:
I think this is the only thing that people should need to read. comments are easily outdated. if the rule is too complicated to read we should either decompose them or come up with some kind of DSL which allows us to specify rules more easily

Reply:
Currently even our simple pattern matching rules take way too long to read for a quick "is this what I need?". Take `fuseReduceMap`, the name does not give me much and the implementation takes way too much effort to read.